### PR TITLE
agent: fix several data races and bugs related to node-local alias checks

### PR DESCRIPTION
**Pull request by rboyer**
The observed bug was that a full restart of a consul datacenter (servers and clients) in conjunction with a restart of a connect-flavored application with bring-your-own-service-registration logic would very frequently cause the envoy sidecar service check to never reflect the aliased service.

Over the course of investigation several bugs and unfortunate interactions were corrected:

## (1) data races

`local.CheckState` objects were only shallow copied, but the key piece of data that gets read and updated is one of the things not copied (the underlying Check with a Status field). When the stock code was run with the race detector enabled this high-relevant-to-the-test-scenario field
was found to be racy.

### Changes

* Update the existing Clone method to include the Check field.
* Use copy-on-write when those fields need to change rather than incrementally updating them in place.

This made the observed behavior occur slightly less often.

## (2) control loop corrections

If anything about how the `runLocal()` method for node-local alias check logic was ever flawed, there was no fallback option. Those checks are purely edge-triggered and failure to properly notice a single edge transition would leave the alias check incorrect until the next flap of the aliased check.

### Change

The change was to introduce a fallback timer to act as a control loop to double check the alias check matches the aliased check every minute (borrowing the duration from the non-local alias check logic body).

This made the observed behavior eventually go away when it did occur.

## (3) overlapping alias check lifetimes

Originally I thought there were two main actions involved in the data race:

**(A)** The act of adding the original check (from disk recovery) and its first health evaluation.

**(B)** The act of the HTTP API requests coming in and resetting the local state when re-registering the same services and checks.

It took awhile for me to realize that there's a third action at work:

**(C)** The goroutines associated with the original check and the later checks.

The actual sequence of actions that was causing the bad behavior was that the API actions result in the original check to be removed and re-added _without waiting for the original goroutine to terminate_. This means for brief windows of time during check definition edits there are
two goroutines that can be sending updates for the alias check status.

In extremely unlikely scenarios the original goroutine sees the aliased check start up in `critical` before being removed but does not get the notification about the nearly immediate update of that check to
`passing`.

This is interlaced with the new goroutine coming up, initializing its base case to `passing` from the current state and then listening for new notifications of edge triggers.

If the original goroutine "finishes" its update, it then commits one more write into the local state of `critical` and exits leaving the alias check no longer reflecting the underlying check.

### Change

The correction here is to enforce that the old goroutines must terminate before spawning the new one for alias checks using `sync.WaitGroup`

### CODE REVIEW

File: agent/testagent.go [hunk 1]
```diff
@@ -126,14 +126,15 @@ func (a *TestAgent) Start(t *testing.T) *TestAgent {
...
-	id := NodeID()
 
+	var id string
 	for i := 10; i >= 0; i-- {
 		a.Config = TestConfig(
 			randomPortsSource(a.UseTLS),
 			config.Source{Name: a.Name, Format: "hcl", Data: a.HCL},
 			config.Source{Name: a.Name + ".data_dir", Format: "hcl", Data: hclDataDir},
 		)
+		id = string(a.Config.NodeID)
```
Comment (rboyer) [inline agent/testagent.go]: This was just a drive-by change that helps in the event that you wish to retain the nodeID between restarts during tests.

File: agent/checks/alias.go [hunk 2]
```diff
@@ -203,6 +229,8 @@ func (c *CheckAlias) processChecks(checks []*structs.HealthCheck) {
 		msg = "All checks passing."
 	}
 
+	// TODO(rb): if no matching checks found should this default to critical?
```
Comment (freddygv) [inline agent/checks/alias.go]: It seems like it should. No matching checks means that there were no registered checks for the aliased service nor for the node.

Comment (rboyer) [reply agent/checks/alias.go]: I left this as a drive-by comment mostly as an observation. I am not sure what the original intended operation of this default behavior was supposed to be and this surprised me.

File: agent/local/state.go [hunk 3]
```diff
@@ -590,6 +596,13 @@ func (l *State) UpdateCheck(id types.CheckID, status, output string) {
 		return
 	}
 
+	// Ensure we only mutate a copy of the check state and put the finalized
+	// version into the checks map when complete.
+	c = c.Clone()
+	defer func(c *CheckState) {
+		l.checks[id] = c
+	}(c)
```
Comment (banks) [inline agent/local/state.go]: We are relying on the defer to run before the deferred unlock. That's fine per Go's spec, but I wonder if we should make it more obvious in a comment here or at the top that if someone ever removes or moves the defer unlock it would be racy?

File: agent/structs/structs.go [hunk 4]
```diff
@@ -1045,6 +1059,8 @@ func (c *HealthCheck) IsSame(other *HealthCheck) bool {
 func (c *HealthCheck) Clone() *HealthCheck {
 	clone := new(HealthCheck)
 	*clone = *c
+	clone.ServiceTags = cloneStringSlice(c.ServiceTags)
+	clone.Definition = *c.Definition.Clone()
```
Comment (banks) [inline agent/structs/structs.go]: I wonder if this get's called in any super hot paths that could negatively affect performance? For example do we ever call this during service discovery (/health) calls?

**Comment by rboyer** [comment]
> I wonder if we need to explain in comments somewhere why the "edge trigger" mechanisms used here doesn't actually prevent all races - it was assumed it was robust and copied in a few places IIRC (leaf cert cache type is one of them maybe).
> 
> I tried to work through a sequence diagram of how the race could happen but I couldn't quite figure it out assuming we only send the trigger event strictly after making the actual update. Did I just miss something or is the real bug here that we edge trigger before the update is actually visible from the source? Any interleaving I explored still ended up with the right value unless that assumption was broken.

We hashed this out off-issue, but summarizing here for the record:

The pattern of having a notification channel and then independently fetching the notifications works fine. In this specific situation during the Check lifecycle:

1. trigger old alias check to stop
2. remove old alias check from notification map
3. create new alias check and install in notification map
4. start new alias check
5. new alias check does initialization and waits for notifications

The problem was that any given named check can have *one* notification channel registered to receive notifications. When (1) causes the old check to continue running until after (5) completes we run into issues where the overall check status may be updated like:

1. upstream check goes critical (as part of it independently being restarted and reinitialized)
2. old alias check is notified
3. remove old check from notification map
4. upstream check is passing; sends notification nowhere
5. new alias check spins up and registers into map
6. new alias check initializes to "passing" after querying for upstream (not via notification)
7. old alias check yolo's into the situation and finally gets around to applying that "critical" update to the alias check
8. now the alias check value is "critical" but the upstream value is "passing" and no more notifications will be sent.

File: agent/structs/structs.go (hunk 4 above)
Comment (rboyer) [reply agent/structs/structs.go]: I removed the changes in `agent/structs` and everything passed the race detector so these fields being deep copied was overkill on my part.

File: agent/local/state.go (hunk 3 above)
Comment (rboyer) [reply agent/local/state.go]: Left a note.

File: agent/checks/alias.go (hunk 2 above)
Comment (rboyer) [reply agent/checks/alias.go]: I'm curious if @banks has any thoughts on this specific point?

Comment (banks) [reply agent/checks/alias.go]: I'd argue not. The current behaviour if you have a service registered with zero checks is that it *will* show up in DNS and service discovery results. If we made an alias Critical if the target service has no matching checks then the target would be considered passing while the alias would be critical which is not what I'd expect.

The overall goal is that the Alias check instance's status should always match the overall aggregate status of the targer service instance which in the case of no checks is implicitly passing.

Comment (rboyer) [reply agent/checks/alias.go]: ah right.
